src/core/lib/iomgr/work_serializer.cc

```diff
@@ -58,14 +59,17 @@ void WorkSerializer::WorkSerializerImpl::Run(
   const size_t prev_size = size_.fetch_add(1);
   // The work serializer should not have been orphaned.
   GPR_DEBUG_ASSERT(prev_size > 0);
-  if (prev_size == 1) {
-    // There is no other closure executing right now on this work serializer.
-    // Execute this closure immediately.
+  // If there is no other callback executing right now on this work serializer,
+  // try to grab the lock and execute the callback immediately.
+  bool expected = false;
+  if (prev_size == 1 && draining_.compare_exchange_strong(expected, true)) {
     if (GRPC_TRACE_FLAG_ENABLED(grpc_work_serializer_trace)) {
       gpr_log(GPR_INFO, "  Executing immediately");
     }
     callback();
-    // Loan this thread to the work serializer thread and drain the queue.
+    size_.fetch_sub(1);
+    draining_.store(false);
```

Comment:
All ops on `size_` and `draining_` are `seq_cst` in this PR. Can you please add some documents on why it's needed and whether the ordering of these fields matter for the correctness of the code?

Reply:
You are right. I think `ack_rel` works here.